### PR TITLE
fix(vault): vault try should avoid using semaphore in non-yieldable phases

### PR DESCRIPTION
**Pull request by windmgc**
### Summary

Vault's `try` function now uses `with_coroutine_mutex` concurrency control, but it may be triggered in phases(for example, init phase during config dereference)that does not support `semaphore:wait`, so problem occurs.

### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been added to `CHANGELOG/unreleased/kong` or adding `skip-changelog` label on PR if unnecessary. [README.md](https://github.com/Kong/kong/blob/master/CHANGELOG/README.md)
- [na] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

**Comment by bungle** [comment]
@windmgc I think this is a good catch. I think the problem might be here:
https://github.com/Kong/kong/blob/master/kong/pdk/vault.lua#L1126-L1129

I think we just need to make a tool that checks if you are running in yieldable context:
https://github.com/Kong/kong/blob/master/kong/tools/utils.lua#L1655-L1664

Different functions may have different limitations. That is where Nginx native ngx.sleep does work ^. So I don't want a yet another `if phase == "init"`  as it is not all that. Other phases have similar problems.

**Comment by windmgc** [comment]
@bungle  Got it!

**Comment by windmgc** [comment]
[The doc](https://github.com/openresty/lua-resty-core/blob/2e2b2adaa61719972fe4275fa4c3585daa0dcd84/lib/ngx/semaphore.md#wait) only mentioned 4 phases but I think other yield-able phases should support wait as well, see the following check in C code:
https://github.com/openresty/lua-nginx-module/blob/c89469e920713d17d703a5f3736c9335edac22bf/src/ngx_http_lua_semaphore.c#L382

**Comment by windmgc** [comment]
@bungle Could you check if this implementation is the thing that you want? Thanks!

File: kong/concurrency.lua [hunk 1]
```diff
@@ -91,7 +89,7 @@ function concurrency.with_coroutine_mutex(opts, fn)
     error("invalid value for opts.on_timeout", 2)
   end
 
-  if get_phase() == "init_worker" then
+  if not check_phase_yieldable() then
```
Comment (hanshuebner) [inline kong/concurrency.lua:92]: `in_yieldable_phase()` would be clearer.  I understand that there is an optional `phase` argument, but as we're not going to supply that in most cases, I'd rather optimize the name for the standard use case.

**Review by hanshuebner — CHANGES_REQUESTED** [review]

File: kong/concurrency.lua (hunk 1 above)
Comment (windmgc) [reply kong/concurrency.lua]: Function renamed!

**Review by hanshuebner — APPROVED** [review]

**Pull request merged**